docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+---
+title: Runtime Hooks for Add-on Management
+authors:
+  - "@killianmuldoon"
+  - "@ykakarap"
+reviewers:
+  - "@vincepri"
+  - "@CecileRobertMichon"
+  - "@enxebre"
+  - "@fabriziopandini"
+  - "@sbueringer"
+creation-date: 2022-04-14
+last-updated: 2022-04-14
+status: implementable
+replaces:
+see-also:
+superseded-by:
+---
+
+# Runtime hooks for Add-on Management
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Glossary](#glossary)
+* [Summary](#summary)
+* [Motivation](#motivation)
+	* [Goals](#goals)
+	* [Non-goals](#non-goals)
+    * [Future work](#futurework)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Runtime hook definitions](#runtime-hook-definitions)
+      * [Before Cluster Create](#before-cluster-create)
+      * [After Control Plane Initialized](#after-control-plane-initialized)
+      * [Before Cluster Upgrade](#before-cluster-upgrade)
+      * [After Control Plane Upgrade](#after-control-plane-upgrade)
+      * [After Cluster Upgrade](#after-cluster-upgrade)
+      * [Before Cluster Delete](#before-cluster-delete)
+    * [Runtime Extensions developer guide](#runtime-extensions-developer-guide)
+    * [Security Model](#security-model)
+    * [Risks and Mitigations](#risks-and-mitigations)
+      * [Runtime Extension blocks Cluster lifecycle indefinitely](#runtime-extension-blocks-cluster-lifecycle-indefinitely)
+* [Alternatives](#alternatives)
+  * [External components watching CAPI resources without hooks](#external-components-watching-capi-resources-without-hooks)
+  * [OpenAPI spec implementation alternatives](#openapi-spec-implementation-alternatives)
+    * [Adding Cluster info to request vs providing only the Cluster name](#adding-cluster-info-to-request-vs-providing-only-the-cluster-name)
+    * [Embedding CAPI types in request vs using runtime.RawExtension](#embedding-capi-types-in-request-vs-using-runtimerawextension)
+* [Upgrade Strategy](#upgrade-strategy)
+    * [Cluster API version upgrade](#clusterapi-version-upgrade)
+    * [Kubernetes version upgrade](#kubernetes-version-upgrade)
+* [Additional Details](#additional-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Version Skew Strategy](#version-skew-strategy)
+* [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
```

Comment:
Might be worth to define what we intend as "add-on" in this proposal. Add-on is a very overloaded word in k8s that can mean different things depending on who you talk to...

Reply:
In the [Cluster API Add-Ons Orchestration](https://docs.google.com/document/d/1TdbfXC2_Hhg0mH7-7hXcT1Gg8h6oXKrKbnJqbpFFvjw/edit#) proposal it's defined as:

> Add-on: an application that extends the functionality of Kubernetes.

Sounds very generic, but I assume it's good enough for the purpose of the current proposal